### PR TITLE
Disallow v0 transactions as a consensus rule

### LINKED ISSUE #1557: version 2 transactions should be standard
At https://github.com/zcash/zcash/blob/v1.0.0-rc1/src/main.cpp#L640 we have:
```
bool IsStandardTx(const CTransaction& tx, string& reason)
{
    if (tx.nVersion > CTransaction::CURRENT_VERSION || tx.nVersion < 1) {
        reason = "version";
        return false;
    }
    ...
```
and in https://github.com/zcash/zcash/blob/v1.0.0-rc1/src/primitives/transaction.h#L297:
```
    static const int32_t CURRENT_VERSION=1;
```

So version 2 transactions are not standard. This is bad! How did transactions with JoinSplits ever get mined or relayed?!

### PR DESCRIPTION
**Pull request by str4d**
Closes #1557

### CODE REVIEW

**Comment by bitcartel** [comment]
utACK

**Comment by str4d** [comment]
Rebased on master so that I could base the block version consensus change off this branch to avoid merge conflicts.

File: src/main.cpp [hunk 1]
```diff
@@ -851,6 +851,12 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
 {
     // Basic checks that don't depend on any context
 
+    // Check transaction version
+    if (tx.nVersion < MIN_TX_VERSION) {
+        return state.DoS(100, error("CheckTransaction(): version too low"),
+                         REJECT_INVALID, "bad-version-too-low");
```
Comment (daira) [inline src/main.cpp]: bad-txns-version-too-low

Comment (daira) [inline src/main.cpp]: I'll tack this change onto the end of #1603 to avoid disruption.

**Comment by daira** [comment]
it(ACK+cov). @zkbot r+

**Comment by daira** [comment]
s/it/ut/

**Comment by str4d** [comment]
@zkbot force

**Comment by ebfull** [comment]
@zkbot r+

**Comment by str4d** [comment]
Okay, the problem is this line in `sighash_test` generation:
```
void static RandomTransaction(CMutableTransaction &tx, bool fSingle) {
    tx.nVersion = insecure_rand();
```

So there are random transactions that have `nVersion = 0`. There are two options:

- Regenerate the tests to only use `nVersion >= 1`
- Modify the tests so that if `nVersion < MIN_TX_VERSION`, the test must fail.

I prefer the latter.

**Comment by daira** [comment]
Yes, I think the latter is more in the spirit of the test.

**Comment by str4d** [comment]
@daira test fixed.

**Comment by daira** [comment]
ACK. @zkbot r+

**Pull request merged**